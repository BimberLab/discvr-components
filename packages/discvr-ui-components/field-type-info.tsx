import React from 'react';
import TextField from '@mui/material/TextField';
import Select from '@mui/material/Select';
import AsyncSelect from 'react-select/async';
import MenuItem from '@mui/material/MenuItem';
import FormControl from '@mui/material/FormControl';
import InputLabel from '@mui/material/InputLabel';
import { styled } from '@mui/material/styles';



interface FieldModel {
    name: string;
    label: string;
    description: string;
    type: string;
    isInDefaultColumns: boolean;
    isMultiValued: boolean;
    isHidden: boolean;
    colWidth: number;
    formatString: string;
    orderKey: number;
    allowableValues: string[];
    category: string;
    url: string;
    flex: number;
    supportsFilter: boolean;
}


type HandleFilterChangeFunction = (
    index: number, 
    key: string, 
    value: string, 
) => FilterType;


interface FilterType {
    field: string;
    operator: string;
    value: string;
}


interface ValueInputProps {
    filter: FilterType;
    index: number;
    highlightedInputs: { field:boolean, operator: boolean, value:boolean }[];
    handleFilterChange: HandleFilterChangeFunction;
    allowedGroupNames?: string[];
    fieldTypeInfo: FieldModel[];
}



function ValueComponent(props: ValueInputProps): React.ReactElement | null {

    const {filter, index, highlightedInputs, handleFilterChange, allowedGroupNames, fieldTypeInfo} = props

    const highlightedSx = {
        border: '2px solid red',
        borderRadius: '4px'
    }

    const FormControlMinWidth = styled(FormControl)(({ theme }) => ({
        minWidth: 200,
        marginRight: theme.spacing(2)
    }))
    
    const TextFieldMinWidth = styled(TextField)(({ theme }) => ({
        minWidth: 200,
        marginRight: theme.spacing(2)
    }))

    return (
        <>
            {
                filter.operator === "in set" ? (
                    <FormControlMinWidth sx={highlightedInputs[index]?.value ? highlightedSx : null} >
                        <InputLabel id="value-select-label">Value</InputLabel>
                        <Select
                            labelId="value-select-label"
                            onChange={(event) =>
                                { handleFilterChange(index, "value", event.target.value); }
                            }
                            value={filter.value}
                        >
                            {allowedGroupNames?.map((gn) => (
                                <MenuItem key={gn} value={gn}>{gn}</MenuItem>
                            ))}
                        </Select>
                    </FormControlMinWidth>
                ) : fieldTypeInfo.find(obj => obj.name === filter.field)?.allowableValues.length > 10 ? (
                    <FormControlMinWidth sx={highlightedInputs[index]?.value ? highlightedSx : null} >
                        <AsyncSelect
                            aria-labelledby={`value-select-${index}`}
                            id={`value-select-${index}`}
                            inputId={`value-select-${index}`}
                            isDisabled={filter.operator === "is empty" || filter.operator === "is not empty"}
                            isMulti={fieldTypeInfo.find(obj => obj.name === filter.field)?.isMultiValued}
                            loadOptions={(inputValue, callback) => {
                                const fieldInfo = fieldTypeInfo.find(obj => obj.name === filter.field);

                                callback(
                                    (fieldInfo?.allowableValues || [])
                                        .filter(value => value.toLowerCase().includes(inputValue.toLowerCase()))
                                        .map(value => ({ label: value, value }))
                                );
                            }}
                            menuPortalTarget={document.body}
                            menuPosition="fixed"
                            menuShouldBlockScroll
                            onChange={(selected: FilterType[]) => { handleFilterChange(index, "value", selected.length > 0 ? selected.map(s => s.value).join(',') : undefined); }}
                            styles={{ menuPortal: base => ({ ...base, zIndex: 9999 }) }}
                            value={filter.value ? filter.value.split(',').map(value => ({ label: value, value })) : undefined}
                        />
                    </FormControlMinWidth>
                ) : fieldTypeInfo.find(obj => obj.name === filter.field)?.allowableValues.length > 0 ? (
                    <FormControlMinWidth sx={highlightedInputs[index]?.value ? highlightedSx : null} >
                        <InputLabel id="value-select-label">Value</InputLabel>
                        <Select
                            disabled={filter.operator === "is empty" || filter.operator === "is not empty"}
                            labelId="value-select-label"
                            onChange={(event) =>
                                { handleFilterChange(index, "value", event.target.value); }
                            }
                            value={filter.value}
                        >
                            {fieldTypeInfo.find(obj => obj.name === filter.field)?.allowableValues.map(allowableValue => (
                                <MenuItem key={allowableValue} value={allowableValue}>
                                    {allowableValue}
                                </MenuItem>
                            ))}
                        </Select>
                    </FormControlMinWidth>
                ) : (
                    <TextFieldMinWidth
                        disabled={filter.operator === "is empty" || filter.operator === "is not empty"}
                        label="Value"
                        onChange={(event) =>
                            { handleFilterChange(index, 'value', event.target.value); }
                        }
                        sx={highlightedInputs[index]?.value ? highlightedSx : null}
                        value={filter.value}
                    />
                )
            }
        </>
    );
}

export { ValueComponent }
