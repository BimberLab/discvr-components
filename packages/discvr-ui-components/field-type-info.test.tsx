import React from 'react';
import {create} from 'react-test-renderer';
import { ValueComponent } from './field-type-info';

export declare interface FieldModel {
    name: string
    label: string
    description: string
    type: string
    isInDefaultColumns: boolean
    isMultiValued: boolean
    isHidden: boolean
    colWidth: number
    formatString: string
    orderKey: number
    allowableValues: string[]
    category: string
    url: string
    flex: number
    supportsFilter: boolean
}

export declare interface FilterModel {
    field: string;
    operator: string;
    value: string;
}

it('ValueComponent test', () => {
    const fieldTypeInfo:FieldModel[] = [{
        name: '',
        label: '',
        description: '',
        type: '',
        isInDefaultColumns: false,
        isMultiValued: false,
        isHidden: false,
        colWidth: 0,
        formatString: '',
        orderKey: 0,
        allowableValues: [],
        category: '',
        url: '',
        flex: 0,
        supportsFilter: false
    },
    {
        name: '',
        label: '',
        description: '',
        type: '',
        isInDefaultColumns: false,
        isMultiValued: false,
        isHidden: false,
        colWidth: 0,
        formatString: '',
        orderKey: 0,
        allowableValues: [],
        category: '',
        url: '',
        flex: 0,
        supportsFilter: false
    }]

    const filter:FilterModel = {
        field: '',
        operator: '',
        value: ''
    }

    


    const component = create(
        <ValueComponent 
        fieldTypeInfo={fieldTypeInfo} 
        filter={filter} 

        handleFilterChange={undefined}

        highlightedInputs={[{field:false, operator:false, value:false},{field:false, operator:false, value:false}]} 
        highlightedSx={undefined} 
        index={0} />
    );
    expect(component.toJSON()).toMatchSnapshot();

});