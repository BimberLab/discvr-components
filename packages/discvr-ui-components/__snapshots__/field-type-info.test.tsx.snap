// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValueComponent test 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root css-1rv32ir-MuiFormControl-root-MuiTextField-root"
>
  <label
    className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={false}
  >
    Value
  </label>
  <div
    className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="text"
      value=""
    />
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-yjsfm1"
      >
        <span>
          Value
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
